### PR TITLE
ErrorHandling.md: Fatal Errors with Promises

### LINKED ISSUE #11: Better specified context events
WebGL has events for loss and restoration of a context and creation failure error. However, there's a problem with them: sometimes they're not very helpful.

For instance, `webglcontextlost` event. Since it doesn't give the reason why we've lost the context, it's not clear what to do: show our user an error message or wait for restore event. I actually tried to log `statusMessage` property of event objects and it was empty.

It'd be good to have better specified context events.

### PR DESCRIPTION
**Pull request by kainino0x**
Adds the most minimal API I could find for handling device/adapter loss cases. Please see the "case studies" section and example code to better understand how this works.

Also fixes #11.

### CODE REVIEW

File: design/RejectedErrorHandling.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+Appendix document for [ErrorHandling.md](ErrorHandling.md).
```
Comment (Kangz) [inline design/RejectedErrorHandling.md]: Note: this is a rejected proposal listed here for posterity.

**Comment by Kangz** [comment]
Thanks for putting this up, it trades synchronicity of create device in favor of making it easy to write robust device and adapter loss handling. Nice!

File: design/ErrorHandling.md [hunk 2]
```diff
@@ -23,6 +23,148 @@ There are several types of WebGPU calls that get their errors handled differentl
 Implementations should provide a way to enable synchronous validation, for example via a debug shim or via the developer tools.
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
+## *Fatal Errors*: Lost/Recovered Events
+
+<!-- calling this revision 5 -->
```
Comment (kvark) [inline design/ErrorHandling.md]: does there need to be revision index in the text?

File: design/RejectedErrorHandling.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    readonly attribute boolean recoverable;
+    readonly attribute DOMString reason;
+};
+```
+
+If `GPUAdapter`'s `isReady` attribute is false, `createDevice` will fail. 
+`isReady` may be set to `false` when a `"gpudevicelost"` event fires.
+It will always be set to `true` when a `"gpuadapterready"` event fires.
+
+ - `GPUAdapter` `"gpuadapterlost" -> GPUAdapterLostEvent`:
```
Comment (kvark) [inline design/RejectedErrorHandling.md]: would it be possible to separate the words with dashes in the reason? i.e. "gpu-adapter-lost"

File: design/ErrorHandling.md (hunk 2 above)
Comment (kainino0x) [reply design/ErrorHandling.md]: I didn't actually save all of the revisions, so I can't really make one, unfortunately.

File: design/RejectedErrorHandling.md (hunk 3 above)
Comment (kainino0x) [reply design/RejectedErrorHandling.md]: This is a rejected revision, so I don't think it's worth updating.

File: design/RejectedErrorHandling.md (hunk 1 above)
Comment (kainino0x) [reply design/RejectedErrorHandling.md]: Added this to the top of the doc

**Comment by beaufortfrancois** [comment]
What do you think of using an event handler instead of this devicelost callback function? Mixing callback functions and promises isn't something I've seen a lot in Web APIs. Note that this proposal would be similar to how [Web Bluetooth](https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothdeviceeventhandlers-ongattserverdisconnected) and [Web USB](https://wicg.github.io/webusb/) handles device disconnections.

See my proposal below 

```webidl
Promise<GPUDevice> requestDevice(GPUDeviceDescriptor descriptor);
```

```webidl
partial interface GPUDevice : EventTarget {
  attribute EventHandler ondevicelost;
};
```

`ondevicelost` is an [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) which is called when something goes fatally wrong on the GPU device (e.g. unexpected out-of-memory, crash, or native device loss).

```webidl
[
 Constructor(DOMString type, GPUDeviceLostEventInit gpuDeviceLostEventInitDict),
 Exposed=Window
]
interface GPUDeviceLostEvent : Event {
  readonly attribute DOMString message;
};

dictionary GPUDeviceLostEventInit : EventInit {
  required DOMString message;
};
```

Fire an event named "devicelost" at GPUDevice using GPUDeviceLostEvent with its message attribute initialized to the error message that triggered device lost.

---

Example would have to be updated as well. Note that `this.device` needs to be reset before and after this change.

```js
    try {
      this.device = await this.adapter.requestDevice({ /* options */ });
      this.device.addEventListener('devicelost', function(event) {
        // Device was lost.
        console.error("device lost", info);
        // Try to get a device again.
        this.ensureDevice();
      });
      this.rendering = true;
    } catch (e) {
```

**Comment by kainino0x** [comment]
For a while before I opened this PR I had it as an event rather than a callback. However I tentatively changed it to a callback and left this TODO:

> (TODO: this could probably be an `ondevicelost` event instead. However, this (1) makes it required and (2) makes it clear that no error can be missed between creation and event listener registration.)

Do you have specific advice?

1. Not necessary for it to be required, it's just nice because it forces developers to notice they need to do something there.
2. I'm not sure whether we can guarantee the event handler will be called if the device is lost before it's set. This seems to work in WebGL today, but I'm not sure if it's guaranteed/guaranteeable:

```js
(() => {
  const cvs = document.createElement('canvas');
  const gl = cvs.getContext('webgl');
  const ext = gl.getExtension('WEBGL_lose_context');
  ext.loseContext();
  cvs.addEventListener('webglcontextlost', (e) => { console.warn(e); });
})();
```

This is necessary because the app has to know it's never going to leave a context loss on the floor, because it won't know to reinitialize.

File: design/ErrorHandling.md [hunk 4]
```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+```webidl
+partial interface GPUAdapter {
+    Promise<GPUDevice> requestDevice(GPUDeviceDescriptor descriptor,
+                                     GPUDeviceLostCallback onDeviceLost);
+};
+
+callback GPUDeviceLostCallback = void (GPUDeviceLostInfo info);
+
+interface GPUDeviceLostInfo {
+    readonly attribute GPUDevice device;
```
Comment (kainino0x) [inline design/ErrorHandling.md]: RC: remove or rename to `lostDevice`

**Comment by beaufortfrancois** [comment]
> 2\. I'm not sure whether we can guarantee the event handler will be called if the device is lost before it's set. This seems to work in WebGL today, but I'm not sure if it's guaranteed/guaranteeable:

In that case, we _could_ have a `lost`boolean attribute added to `GPUDevice` that would reflect the device lost state. This is what does the [Web Bluetooth](https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattserver-connected) spec. Note that [WebUSB](https://wicg.github.io/webusb/#dom-usb-onconnect) doesn't do this.

```webidl
partial interface GPUDevice : EventTarget {
  readonly attribute boolean lost;
  attribute EventHandler ondevicelost;
};
```

**Comment by kainino0x** [comment]
I've been trying hard for there to not be two separate signals the app *has* to check to make sure they do things right, so I'd strongly like to avoid it.

If it were something like `readonly attribute Promise<void> ondevicelost`, and app has to do `device.ondevicelost.then(...)`, it would guarantee that a late-registered handler would be called. But this kind of design would be very unusual, I think.

File: design/ErrorHandling.md [hunk 5]
```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+    readonly attribute DOMString reason;
+};
+```
+
+`GPUAdapter.requestDevice` requests a device from the adapter.
+It returns a Promise which resolves when a device is ready.
+The Promise may not resolve for a long time - it resolves when the browser is ready for the application to bring up (or restore) its content.
+If the adapter is lost before the Promise resolves, the Promise rejects.
```
[the remark this answers is not in the document]
Comment (Kangz) [reply design/ErrorHandling.md]: Wouldn't this remove all possibilities to get data about the device before creating it, for example to know which extensions / limits are present and can be requested?

**Comment by grorg** [comment]
Discussed at the [11 Feb 2019 WebGPU Meeting](https://docs.google.com/document/d/1iLZAb6LWiqjoAxsIORXbYdMpspF5a0tIa2uFgUTc_7c/)

**Comment by kainino0x** [comment]
Discussed requestAdapter vs listAdapters at the meeting. The issue is somewhat orthogonal.

I will consider at Rafael's other comments, try to describe the meaning of requestAdapter rejection, and then we'll try to merge this ASAP so it doesn't collide with requestAdapter vs listAdapters.

File: design/ErrorHandling.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (RafaelCintron) [reply design/ErrorHandling.md]: Yes, @Kangz , it would remove the possibility of getting data about the device before creating it.  The same is true of the WebXR, other media based APIs and even getContext.  For those, you pass information about what the context/device you would like and the browser returns to you a context/device that meets the criteria.

File: design/ErrorHandling.md [hunk 6]
```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+    try {
+      this.device = await this.adapter.requestDevice({ /* options */ }, (info) => {
+        // Device was lost.
+        console.error("device lost", info);
+        // Try to get a device again.
+        this.ensureDevice();
+      });
+      this.rendering = true;
+    } catch (e) {
+      // Request failed (likely due to adapter loss).
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: I actually considered and specifically avoided this additional signal. I don't think there is any case in which the application would need to know about an adapter loss, aside from knowing that the device(s) were lost. They can try adapter.createDevice to discover whether the adapter will no longer vend any new devices.

Also, since one device on an adapter can be lost without other devices on the same adapter being lost, the signal needed to be on the device.

File: design/ErrorHandling.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: To be discussed after this PR.

File: design/ErrorHandling.md [hunk 7]
```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+  }
+  initFallback() { /* ... */ }
+  async ensureDevice() {
+    // Stop rendering. (If there was already a device, WebGPU calls made before
+    // the app notices the device is lost are okay - they are no-ops.)
+    this.rendering = false;
+    if (!this.adapter) {
+      // If no adapter, get one.
+      // (If requestAdapter rejects, no matching adapter is available. Exit to fallback.)
+      this.adapter = await gpu.requestAdapter({ /* options */ });
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: Actually, the comment just above this line explains what happens here: if requestAdapter() rejects, ensureDevice() rejects, and the app falls back to non-webgpu.

File: design/ErrorHandling.md [hunk 8]
```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+    }
+    try {
+      this.device = await this.adapter.requestDevice({ /* options */ }, (info) => {
+        // Device was lost.
+        console.error("device lost", info);
+        // Try to get a device again.
+        this.ensureDevice();
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: This said, the sample code *could* just always try on a newly requested adapter instead of retrying on the same adapter before trying on a new adapter. Both would be valid choices for an app, but the distinction also seems minimal. Maybe it would be better for the sample to be simpler.

[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: Decided not to simplify the example, because it's useful to demonstrate the use case. However, I did refactor the example to be iterative instead of recursive, which is clearer. I think it is equivalent.

**Comment by kainino0x** [comment]
FYI: Updated the example (and rebased), but haven't resolved onDeviceLost vs event vs promise yet. So that part is not ready for review.

**Comment by kainino0x** [comment]
Here is a more comprehensive investigation of the options I've considered so far. @beaufortfrancois and @jungkees in particular, would you be able to comment on this?

Premises:
1. A device will never be lost more than once.
2. An application needs to be able to guarantee it won't miss a device loss happening.
    1. An application probably should not *be able* to write code that *may or may not* miss a device loss. (That is, any code they write should either never miss, or always miss (because the code is wrong).)
4. An application should be able to register a handler that won't collide with other applications on the same page. (It would be common for multiple apps, e.g. chart widgets or model viewers, to exist together in the same page.)
    1. An application probably should not *be able* to easily collide with other applications on the same page.
5. It probably shouldn't mismatch other Web Platform APIs.

Options:
- Event handler registered before device creation (precedent: canvas receives `webglcontextlost`; `navigator.usb` receives `ondisconnect`; `navigator.bluetooth` receives `ongattserverdisconnected`).
  - Doesn't express premise 1.
  - Satisfies premise 3, but not premise 3.i.: Each app on the page requires extra logic to make sure it doesn't collide with others. Since it doesn't happen by default, apps won't be compatible with running multiple-per-page unless they add this extra logic.
  - (Aside: In WebGL, canvas can only have one context, so this works.)
 
```js
let device;
adapter.addEventListener('ondevicelost', (e) => {
  if (e.lostDevice !== device) { return; }
  // ...
});
device = await adapter.createDevice(...);
```

- Event handler on device: registered after device creation (no known precedent).
  - May break premise 2; definitely breaks premise 2.i. Not sure we can guarantee the event is delivered to the handler. Even if we could, it's tricky for the app to make sure it registers the event listener early enough (just inserting an extra await between creation and registration could break it).

```js
const device = await adapter.createDevice(...); // Note that this awaits
device.addEventListener('ondevicelost', (e) => { /* ... */ }); // Not guaranteed?
```

- Provide callback on creation (as an argument to createDevice, or in the device descriptor).
  - Unusual (breaks premise 4).

- Provide event listener on creation (as an argument to createDevice, or in the device descriptor).
  - Unusual (breaks premise 4).
  - Assuming the device is an EventTarget, an application can still accidentally do the wrong thing by doing the addEventListener too late.

- Device has `readonly attribute Promise<void> onlost`. It is always pending until device loss.
  - Unusual (breaks premise 4).
  - Easy to guarantee the event is handled.
  - Could potentially be merged with `getObjectStatus`, if `getObjectStatus` only resolves on invalid (pending forever on valid or on device loss).

```js
const device = await adapter.createDevice(...);
device.onlost.then(() => { /* ... */ }); // Guaranteed, no matter how late it is added.
```

**Comment by RafaelCintron** [comment]
@kainino0x , thank you for the comprehensive investigation.  

Given the fact that a device cannot become lost more than once, I think the promise approach is best because it handles multiple notifications and guarantees notifications will not be missed.  

I'm not concerned about breaking premise 4 because in the other APIs you list, a 'device' can become, restored/reconnected, etc.  Hence, a callback or event is the only viable option because there can be multiple notifications

**Comment by kainino0x** [comment]
I think I forgot to mention above: the first option (webglcontextlost-style) *would* work if requestAdapter always vends a new Adapter object instead of returning the same one for the same physical adapter.

**Comment by jungkees** [comment]
For https://github.com/gpuweb/gpuweb/pull/198#issuecomment-463451548, I think the promise option fits very well. As @RafaelCintron mentioned, it's a state transition rather than an event that can fire multiple times.

I'd like to suggest `.lost` or `.whenLost` instead of `.onlost`. `onlost` sounds like a name of an event handler. The example 4 in https://w3ctag.github.io/promises-guide/#webidl-examples shows that it uses like `stateMachine.loaded` without `on`. It's also different from `.lost()` which is an action. So, `device.lost` sounds fine to me. (https://w3ctag.github.io/promises-guide/#state-transitions has a guide when to use a method and when to use an attribute.)

And actually we have a similar attribute in the platform: https://w3c.github.io/ServiceWorker/#dom-serviceworkercontainer-ready. People initially proposed `.whenReady` but landed on `.ready`.

FYI, there's a patter for option 2 (event handler on device) as well. We can define a [task source](https://html.spec.whatwg.org/#task-source) on the EventTarget (like https://w3c.github.io/ServiceWorker/#dfn-client-message-queue) and queue events before the event handlers or event listeners are attached. We can make the task source hold that events until it is enabled. And we can enable it when the first event handler is attached and/or by addig an API to enable it like https://w3c.github.io/ServiceWorker/#navigator-service-worker-startMessages.

**Comment by kainino0x** [comment]
Thank you very much for the review, and pointer to the prior work with ServiceWorkerContainer.ready and option 2. It'll be helpful to take a look at how those work.

**Comment by kainino0x** [comment]
Applied changes to have `Promise<GPUDeviceLostInfo> lost;` on GPUDevice.

Also changed the example again to remove the loop; ensureDevice doesn't need to try more than twice to get a device, because it always starts from the top on device loss. I think(?) the loop was an artifact of an old intermediate state.

**Comment by kainino0x** [comment]
Changes are not too major; see https://github.com/gpuweb/gpuweb/pull/198/commits/8ecd8bb01d072649f43d913388b3e79c6dd83c2f.

Maybe we can cover and finish this up at the upcoming meeting. Apologies again for the late review request.

File: design/ErrorHandling.md [hunk 9]
```diff
@@ -157,9 +297,6 @@ It is only necessary if an application wishes to recover from recoverable errors
 
 ## Open Questions and Considerations
 
- - WebGPU could guarantee that objects such as `GPUQueue` and `GPUFence` can never be errors.
```
Comment (beaufortfrancois) [inline design/ErrorHandling.md]: Why is is deleted in this PR?

File: design/sketch.webidl [hunk 10]
```diff
@@ -802,7 +807,9 @@ interface GPUAdapter {
     readonly attribute GPUExtensions extensions;
     //readonly attribute GPULimits limits; Don't expose higher limits for now.
 
-    GPUDevice createDevice(GPUDeviceDescriptor descriptor);
+    // May reject with DOMException  // TODO: DOMException("OperationError")?
+    Promise<GPUDevice> requestDevice(GPUDeviceDescriptor descriptor,
+                                     GPUDeviceLostCallback onDeviceLost);
```
Comment (beaufortfrancois) [inline design/sketch.webidl]: We should remove `onDeviceLost` now.

File: design/sketch.webidl [hunk 11]
```diff
@@ -25,6 +25,11 @@ dictionary GPUExtent3D {
 // ERROR HANDLING
 // ****************************************************************************
 
+interface GPUDeviceLostInfo {
+    readonly attribute DOMString reason;
+};
+callback GPUDeviceLostCallback = void (GPUDeviceLostInfo info);
```
Comment (beaufortfrancois) [inline design/sketch.webidl]: Can we add `lost` to the sketch.webidl file?

File: design/ErrorHandling.md [hunk 12]
```diff
@@ -23,6 +23,146 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+
+partial interface GPUDevice {
+    readonly attribute Promise<GPUDeviceLostInfo> lost;
+};
+```
+
+`GPUAdapter.requestDevice` requests a device from the adapter.
+It returns a Promise which resolves when a device is ready.
+The Promise may not resolve for a long time - it resolves when the browser is ready for the application to bring up (or restore) its content.
+If the adapter is lost before the Promise resolves, the Promise rejects.
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: The adapter. If the adapter is lost, it can't create a device. I'll rephrase it to express the intent though.

> If the adapter is unable to create a device (i.e. because the adapter was lost), the Promise rejects.

File: design/ErrorHandling.md [hunk 13]
```diff
@@ -23,6 +23,146 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+
+The device is lost because of an unexpected error in the implementation.
+ - `device.lost` resolves, message = whatever the unexpected thing was.
+
+A TDR-like scenario occurs.
+ - The adapter is lost, which loses all devices on the adapter.
+   `device.lost` resolves on every device, message = adapter reset. Application must request adapter again.
+ - (TODO: alternatively, adapter could be retained, but all devices on it are lost.)
+
+All devices and adapters are lost (except for software?) because GPU access has been disabled by the browser (for this page or globally).
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply design/ErrorHandling.md]: We do, but changing it probably would not cause adapter loss. What I'm referring to here is Chrome's mechanism for guarding against attacks against the GPU process: I don't remember the policy exactly, but roughly speaking, if a page causes a GPU process crash, it doesn't get WebGL access anymore until the user reallows it (or maybe reloads). If the GPU process crashes 3 times, GPU access is disabled across the entire browser until it's restarted.

File: design/ErrorHandling.md (hunk 9 above)
Comment (kainino0x) [reply design/ErrorHandling.md]: Oops, splitting error. This is meant to be in 215.

File: design/sketch.webidl (hunk 11 above)
Comment (kainino0x) [reply design/sketch.webidl]: Oops, forgot to update sketch.

File: design/sketch.webidl (hunk 10 above)
Comment (kainino0x) [reply design/sketch.webidl]: Done

**Comment by kainino0x** [comment]
Thanks for the review!

File: design/ErrorHandling.md [hunk 14]
```diff
@@ -23,6 +23,147 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+    - Both `createDevice` Promises resolve.
+      (Unless the adapter was lost, in which case they would have rejected.)
+
+A page begins loading in a tab, but then the tab is backgrounded.
+ - On load, the page attempts creation of a device.
+    - `createDevice` Promise will resolve.
+
+A device's adapter is physically unplugged from the system (but an integrated GPU is still available).
+ - The same adapter, or a new adapter, is plugged back in.
+    - A later `requestAdapters` call may return the new adapter. TODO: `"gpuadapterschanged"`?
```
Comment (RafaelCintron) [inline design/ErrorHandling.md]: Nit: If you haven't already done so, please open an issue to consider "gpuadapterschanged".  Upon resolution of the issue, the TODO would get removed or replaced with the new plan.

File: design/ErrorHandling.md [hunk 15]
```diff
@@ -23,6 +23,147 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+    } catch (e) {
+      console.error(e);
+      this.initFallback();
+    }
+  }
+  async initWebGPU() {
+    await this.ensureDevice();
+    // ... Upload resources, etc.
+  }
+  initFallback() { /* ... */ }
```
Comment (RafaelCintron) [inline design/ErrorHandling.md]: Nit: Initially, thought this was one of WebGPU sub-fallbacks.  

Recommend changing the comment to say `/* try WebGL or 2D Canvas */`, or something similar.

File: design/ErrorHandling.md [hunk 16]
```diff
@@ -23,6 +23,147 @@ There are several types of WebGPU calls that get their errors handled differentl
...
+```
+
+`GPUAdapter.requestDevice` requests a device from the adapter.
+It returns a Promise which resolves when a device is ready.
+The Promise may not resolve for a long time - it resolves when the browser is ready for the application to bring up (or restore) its content.
+If the adapter is unable to create a device (i.e. because the adapter was lost), the Promise rejects.
+
+The `GPUDevice` may be lost if something goes fatally wrong on the device (e.g. unexpected out-of-memory, crash, or native device loss).
+The `GPUDevice` provides a promise, `device.lost`, which resolves when the device is lost.
+The returned `lost` promise will **never** reject and may be pending forever.
```
Comment (kvark) [inline design/ErrorHandling.md]: it's a bit weird to talk about "returned" when there is no explicit function to return it (i.e. this is an attribute)

**Comment by kainino0x** [comment]
Done. I'm going to go ahead with merge. Thanks all!
